// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";

package Bicache;

// The greeting service definition.
service KV_service{
  // Sends a greeting
  rpc Get(GetRequest) returns (GetReply) {}
  rpc Set(SetRequest) returns (SetReply) {}
//  rpc Mget() returns () {}
//  rpc Set() returns () {}
//  rpc Mset() returns () {}
}

message GetRequest{
  //该节点分配的位置
  uint32 req_id = 1;
  string key = 2;
  uint64 pos_of_key = 3;
  uint64 timestamp = 4;
  //times to extension existence
  bool read_replica = 5;
  uint64 update_times= 6;
}

message GetReply{
  bool is_found = 1;  
  string value = 2;
  uint64 close_pos = 3;
  uint64 timestamp = 4;
  bool node_overloaded = 5;
  uint32 status_code = 255;
}

message SetRequest{
  //该节点分配的位置
  uint32 req_id = 1;
  string key = 2;
  uint64 pos_of_key = 3;
  string value = 4;
  uint64 timestamp = 5;
  //times to extension existence
  uint64 update_times= 6;

}
message SetReply{
  bool is_set = 1;  
  string value = 2;
  uint64 close_pos = 3;
  uint64 timestamp = 4;
  uint32 status_code = 255;
}